picard/plugin.py

```diff
@@ -213,28 +316,124 @@ def load_plugin(self, name, plugindir):
+                update = dirpath or filepaths
+                if plugin_data and plugin_name:
+                    # zipped module from download
+                    zip_plugin = plugin_name + '.zip'
+                    dst = os.path.join(USER_PLUGIN_DIR, zip_plugin)
+                    if update:
+                        dst += '.update'
+                        if os.path.isfile(dst):
+                            os.remove(dst)
+                    ziptmp = tempfile.NamedTemporaryFile(delete=False,
+                                                         dir=USER_PLUGIN_DIR).name
+                    try:
+                        with open(ziptmp, "wb") as zipfile:
+                            zipfile.write(plugin_data)
+                            zipfile.flush()
+                            os.fsync(zipfile.fileno())
+                        os.rename(ziptmp, dst)
+                        log.debug("Plugin saved to %r", dst)
+                    except:
+                        try:
+                            os.remove(ziptmp)
+                        except (IOError, OSError):
+                            pass
+                        raise
+                elif os.path.isfile(path):
+                    dst = os.path.join(USER_PLUGIN_DIR, os.path.basename(path))
+                    if update:
+                        dst += '.update'
+                        if os.path.isfile(dst):
+                            os.remove(dst)
+                    shutil.copy2(path, dst)
+                elif os.path.isdir(path):
+                    dst = os.path.join(USER_PLUGIN_DIR, plugin_name)
+                    if update:
+                        dst += '.update'
+                        if os.path.isdir(dst):
+                            shutil.rmtree(dst)
+                    shutil.copytree(path, dst)
+                if not update:
+                    installed_plugin = self.load_plugin(zip_plugin or plugin_name, USER_PLUGIN_DIR)
+                    if installed_plugin is not None:
+                        self.plugin_installed.emit(installed_plugin, False)
+                else:
+                    self.plugin_updated.emit(plugin_name, False)
             except (OSError, IOError):
                 log.warning("Unable to copy %s to plugin folder %s" % (path, USER_PLUGIN_DIR))
 
+    def query_available_plugins(self, callback=None):
+        self.tagger.xmlws.get(
+            PLUGINS_API['host'],
+            PLUGINS_API['port'],
+            PLUGINS_API['endpoint']['plugins'],
+            partial(self._plugins_json_loaded, callback=callback),
+            xml=False,
+            priority=True,
+            important=True
+        )
+
+    def _plugins_json_loaded(self, response, reply, error, callback=None):
+        if error:
+            self.tagger.window.set_statusbar_message(
+                N_("Error loading plugins list: %(error)s"),
+                {'error': unicode(error)},
```

Comment:
`error` is only a number, `reply.errorString()` returns a textual representation of the error that's useful even for users.